### PR TITLE
Feature branch : Project: Authentication

### PR DESCRIPTION
**Pull request by KossySteve**
A microverse project to work with authentcation in rails, creating posts and restricting non_users from access to view posts.
Please ReadMe file on how to run project

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,66 @@
...
+git clone https://github.com/mimipeshy/re-former.git
+```
+2. Run `cd members_only` to move inside the downloaded repository.
+3. Run `bundle install` to download the project's dependencies.
+4. Run `rails db:migrate` to set up your local database.
+5. Run `rails server` and navigate to http://localhost:3000/users/new in a web browser to view the user creation form.
+6. To register as a user navigate to: http://localhost:3000/users/sign_up.
+7. After registering you can create posts.
+
+
```
Comment (christianotieno) [inline README.md:33]: ##  Improvement  on the commit messages
- [ ] In your next iteration, I would urge to commit as often as posbblie to make it easy to follow the development journey let's aim to make our commits more descriptive ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_commits.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

File: app/controllers/registrations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+
+    private
+    def sign_up_params
+        params.require(:user).permit(:name, :email, :password, :password_confirmation)
+    end
+
+    def update_params
+        params.require(:user).permit(:name, :email, :password, :password_confirmation)
+    end
```
Comment (christianotieno) [inline app/controllers/registrations_controller.rb:12]: ## Implent user sign out feature
- [ ] Now that our user can sign up, we don't have a way of them logging out. Create a method to enable this feature. Make sure  I as User I can successfully sign_out ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_out_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
```
Comment (christianotieno) [inline app/models/user.rb:3]: ## Commented out code
- [ ] Comments: It’s better to delete these **commented pieces of code**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

File: app/views/devise/sessions/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+<h2>Log in</h2>
+
```
Comment (christianotieno) [inline app/views/devise/sessions/new.html.erb:2]: ## Provide  sign in/signup and sign out links to the user
- [ ] Once we have implemented the logic for a user to log out, let's make ensure we have a Sign in/Sign out links present in any our views here ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_in_out_link_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49722686/))

File: app/views/posts/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  <tbody>
+    <% @posts.each do |post| %>
+      <tr>
+      <% if user_signed_in? %>
+        <td><%= post.user.name %></td>
+      <%end%>
+        <td><%= post.post %></td>
+        <td><%# link_to 'Show', post %></td>
+        <td><%# link_to 'Edit', edit_post_path(post) %></td>
+        <td><%# link_to 'Destroy', post, method: :delete, data: { confirm: 'Are you sure?' } %></td>
```
Comment (christianotieno) [inline app/views/posts/index.html.erb]: ## Commented out code
- [ ] Comments: It’s better to delete these **commented pieces of code**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

**Review by christianotieno — CHANGES_REQUESTED** [review]
## Changes Required ♻️ 
Hello team 👋 . Progress is a decent one. 
The project, however, is still undone, kindly take a gape at the propositions and work on the requirements below. 👇 

### Christian Otieno (TSE)

File: app/views/posts/_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    <div id="error_explanation">
+      <h2><%= pluralize(post.errors.count, "error") %> prohibited this post from being saved:</h2>
+
+      <ul>
+        <% post.errors.full_messages.each do |message| %>
+          <li><%= message %></li>
+        <% end %>
+      </ul>
+    </div>
+  <% end %>
```
Comment (BrittanyBlake) [inline app/views/posts/_form.html.erb:12]: - View-related logic should be encapsulated in View helpers `(app/helpers/*_helper.rb)`

File: app/views/posts/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  </thead>
+
+  <tbody>
+    <% @posts.each do |post| %>
+      <tr>
+      
+        <td><%= post.post %></td>
+        <% if user_signed_in? %>
+        <td>from <%= post.user.name %></td>
+        <%end%>
```
Comment (BrittanyBlake) [inline app/views/posts/index.html.erb:21]: - View-related logic should be encapsulated in View helpers `(app/helpers/*_helper.rb)`

**Review by BrittanyBlake — APPROVED** [review]
### Status: Approved ✅ 
****
### Hi Team! 👋 
### Great job implementing the requested changes!! 🎖️ It's a 👍 from me! Go ahead and merge this pull request and keep up the good work 😄
****
**[OPTIONAL]**

- [ ] There should be **absolutely no logic in your views**. All view related logic should be encapsulated in view helpers `app/helpers/*_helper.rb`.  
    - As this was not pointed out to you in your previous review I have marked it as optional but I highly recommend making these changes as it is a requirement and it is good practice. 
   - I have added comments below of the logic in your views that should be in helpers.

_P.S.  all requested changes marked by [OPTIONAL] are not required. It is up to you whether or not you would like to do them_ 😄 

 Good Luck and Happy Coding! 🍀 👩🏾‍💻

**Pull request merged**